lightning/src/ln/wire.rs

```diff
@@ -0,0 +1,411 @@
+			Ok(Message::Init(Readable::read(buffer)?))
+		},
+		msgs::ErrorMessage::TYPE => {
+			Ok(Message::Error(Readable::read(buffer)?))
+		},
+		msgs::Ping::TYPE => {
+			Ok(Message::Ping(Readable::read(buffer)?))
+		},
+		msgs::Pong::TYPE => {
+			Ok(Message::Pong(Readable::read(buffer)?))
+		},
+		msgs::OpenChannel::TYPE => {
+			Ok(Message::OpenChannel(Readable::read(buffer)?))
+		},
+		msgs::AcceptChannel::TYPE => {
+			Ok(Message::AcceptChannel(Readable::read(buffer)?))
+		},
+		msgs::FundingCreated::TYPE => {
+			Ok(Message::FundingCreated(Readable::read(buffer)?))
+		},
+		msgs::FundingSigned::TYPE => {
+			Ok(Message::FundingSigned(Readable::read(buffer)?))
+		},
+		msgs::FundingLocked::TYPE => {
+			Ok(Message::FundingLocked(Readable::read(buffer)?))
+		},
+		msgs::Shutdown::TYPE => {
+			Ok(Message::Shutdown(Readable::read(buffer)?))
+		},
+		msgs::ClosingSigned::TYPE => {
+			Ok(Message::ClosingSigned(Readable::read(buffer)?))
+		},
+		msgs::UpdateAddHTLC::TYPE => {
+			Ok(Message::UpdateAddHTLC(Readable::read(buffer)?))
+		},
+		msgs::UpdateFulfillHTLC::TYPE => {
+			Ok(Message::UpdateFulfillHTLC(Readable::read(buffer)?))
+		},
+		msgs::UpdateFailHTLC::TYPE => {
+			Ok(Message::UpdateFailHTLC(Readable::read(buffer)?))
+		},
+		msgs::UpdateFailMalformedHTLC::TYPE => {
+			Ok(Message::UpdateFailMalformedHTLC(Readable::read(buffer)?))
+		},
+		msgs::CommitmentSigned::TYPE => {
+			Ok(Message::CommitmentSigned(Readable::read(buffer)?))
+		},
+		msgs::RevokeAndACK::TYPE => {
+			Ok(Message::RevokeAndACK(Readable::read(buffer)?))
+		},
+		msgs::UpdateFee::TYPE => {
+			Ok(Message::UpdateFee(Readable::read(buffer)?))
+		},
+		msgs::ChannelReestablish::TYPE => {
+			Ok(Message::ChannelReestablish(Readable::read(buffer)?))
+		},
+		msgs::AnnouncementSignatures::TYPE => {
+			Ok(Message::AnnouncementSignatures(Readable::read(buffer)?))
+		},
+		msgs::ChannelAnnouncement::TYPE => {
+			Ok(Message::ChannelAnnouncement(Readable::read(buffer)?))
+		},
+		msgs::NodeAnnouncement::TYPE => {
+			Ok(Message::NodeAnnouncement(Readable::read(buffer)?))
+		},
+		msgs::ChannelUpdate::TYPE => {
+			Ok(Message::ChannelUpdate(Readable::read(buffer)?))
+		},
+		_ => {
+			Ok(Message::Unknown(MessageType { number: message_type }))
+		},
+	}
+}
+
+/// Writes a message to the data buffer encoded as a 2-byte big-endian type and a variable-length
+/// payload.
+///
+/// # Errors
+///
+/// Returns an I/O error if the write could not be completed.
+pub fn write<E: Encode, W: Writer>(message: &E, buffer: &mut W) -> Result<(), ::std::io::Error> {
+	<E as Encode>::write(&message, buffer)
+}
+
+/// Defines a type-identified encoding for sending messages over the wire.
+///
+/// Messages implementing this trait specify a type and must be [`Writeable`]. The default [`write`]
+/// implementation handles writing the message type and payload.
+///
+/// [`Writeable`]: ../../util/ser/trait.Writeable.html
+/// [`write`]: #method.write.html
+pub trait Encode: Writeable {
```

Comment:
Hmm, can we name it something a bit more descriptive? Like just call it Message so it can be referred to as wire::Message or WireMessage?

Reply:
Hmmm...So I find the current text confusing, and I dont think sticking to someone elses' style guide a compelling reason to have something that reads confusing. In general, for our own internal traits, I don't really see much reason why we should care in the slightest what some rust community members think traits should be named, at least unless we have no other criteria for determining (eg, readability).

Reply:
I would disagree with Matt regarding sticking to other people's style conventions. I think it makes sense to stick to commonly agreed upon conventions such that our code becomes more easy to read and modify. Given the purpose of this library is extensibility, external readers should not be confused by names, and e. g. traits should be clearly named as verbs, like in other Rust code, too.

Reply:
I don't mind it as a last-resort naming criteria, but readability should always override any style conventions (considering that the conventions are there for readability to begin with), IMO.

Reply:
Is there a project convention for naming traits?

I can think of two reasonable names in replace of `Encode`, but the readability of each depends on the context.

`Encodable` (_adjective_) reads well in trait bounds but not trait in implementations.

```rust
pub fn write<E: Encodable, W: Writer>(message: &E, buffer: &mut W) ->Result<(), ::std::io::Error> {}
impl Encodable for msgs::Ping {}
```

`Encoding` (_noun_) reads well in trait implementations but not in trait bounds.

```rust
pub fn write<E: Encoding, W: Writer>(message: &E, buffer: &mut W) ->Result<(), ::std::io::Error> {}
impl Encoding for msgs::Ping {}
```

Rust guides typically refer to traits as behaviors (verbs) rather than characteristics (adjectives). In this particular instance, we're defining a single behavior. And, as such, a verb seems more natural to me. Nouns seem more useful when talking about traits defining a set of behaviors.

@TheBlueMatt You objected to `Encode` by stating that it is confusing. Could you clarify _what_ is confusing about it? It would help in determining a suitable name.

You did mentioned readability but I think that depends on (a) the context as mentioned above and (b) the way in which you read a trait bound. For the latter, you can read the colon in a trait bound as either "is" or "has trait", each of which effects the readability differently. My feeling is when reading as "has trait" the verb usage reads fine. Perhaps there are other readings that I'm not thinking of, though.

Reply:
Oops lol, I guess that wasnt' clear at all. My issue is that it is confusing as it gives no context. This is Encodeable *messages*, as we already have various encoding traits which indicate something different.

Reply:
Ah, ha! Given that the `wire` module is for encoding/decoding messages over the wire, the context within the file should be clear. If needed outside the file, a clearer name could be given using import renaming.

Reply:
Right, it is, however, pub, so we should be careful to make sure it has a clear name in any context, IMO.

Reply:
Chatted offline. Keeping it as `Encode` for now but I've decoupled the trait from `Writeable`. Instead, the module's `write` function will take a message bounded by both traits. We can revisit the naming in conjunction with `Writeable` in the future.